### DISCUSSION TITLE
Improvements in Tabs

### DISCUSSION THREAD

**Discussion by Sparkenstein** [mantinedev/mantine#1810] [Feature requests]
I am working on a desktop application, using Mantine for that.
I need tabs to behave like they would in a browser, + button at the end of all tabs,
middle click closes tabs etc. I have so far achieved the result with few hacks, like
to create a button at the end you need to provide `button` as `label` otherwise `onClick`
doesn't get registered.

It'd be great if the tabs support few other features, there are many things that are doable. 
listing them down here:
1. UI variants: soft-rounded (pill like), closable (with x in built)
2. Tabs alignment: left, center right
3. MOST IMP: lazily mounted tabs and keepMounted behaviour (will explain below)
4. maybe a useTabs hook to custom implement entire tabs? 
5. controllable tabs overflow behavior: currently overflow just wraps because of flexWrap

About lazy loaded tabs, currently if you switch tabs, the content from previous tabs resets, all states in that
component are lost. it's really important for my application to keep the mounted tabs as they were and not loose
states. let me know what do you think about this. None of the other feature matters to me at the moment 
because I was able to achieve rest with multiple hacks. but keepMounted is something I don't I can achieve with
current tabs implementation in mantine

**Comment by Sparkenstein** [comment]
omg what a coincidence, was just looking at other discussions and found this: https://v5.mantine.dev/changelog/5-0-0/

looks like tabs won't get unmounted by default now. I am happy with it
 but wouldn't it be better if this was a prop instead of a default behavior

**Reply by pkro** [reply]
I would like this to be a optional, too, as I have multiple panels in an admin page where I would like to reload data from the API whenever the tab is activated (data posted from tabs can influence the displayed data in others, and I want the API always to be the source of truth and not escalate it into an application state). It would be fine if the default behavior is "no unmount" as I guess it makes sense for the majority of use cases. Maybe it can be accomplished with onTabChange and useEffect inside the components, but that would mean changing many components and would probably re-render all components that were once mounted at least once on every tab change. This keeps me from upgrading to 5 for now.

**Reply by Sparkenstein** [reply]
I spoke with the author on discord. they are not interested in this feature. they said "implement this on user side"

**Reply by pkro** [reply]
thanks for the heads up!

**Reply by baba43** [reply]
That's a real shame and doesn't make sense in my eyes either. Tabs can be used for everything, not only to display small blocks of text.

We use tabs sometimes as a kind of "mini navigation" (also in an administration) and behind each tab there are sometimes smaller, sometimes more complex components. In this respect, it makes no sense at all to load components that are not used. 

That brings me to the question: do we abuse tabs or did the authors of this change simply not consider this use case? 

My next question would be, how are we supposed to fix this problem in a meaningful way? Can a solution simply be implemented through the APIs or run this through controlled state? If it is implemented this way, are there any possible drawbacks / problems regarding the other functionality? 

In my opinion, this use case should be explained in the documentation in any case, so that not everyone develops his own quick and dirty solution that ends up being incompatible with Mantine.

**Reply by rtivital** [reply]
There aren't any problems to be fixed. Tabs component implements [WAI-ARIA Tabs pattern](https://www.w3.org/WAI/ARIA/apg/example-index/tabs/tabs-manual.html). The component is meant to work exactly like it is working right now. Additionally, the component does not restrict you in any way – you can add any additional elements, control state, customize behavior and change styles (all of these features are shown in documentation).

**Reply by baba43** [reply]
@rtivital thanks for posting here. Does this argument also refer to the fact that all tabs are always mounted or is this the necessary condition for the rest of the functionality? Because I'm still sure that everyone who uses tabs as navigation doesn't want every page to be mounted by default, as this just doesn't fit the usual React workflow (triggering cascades even though they're not needed).

**Reply by pkro** [reply]
So here's a quick way to force mount/unmount on tab change using controlled tabs:

    // ...
    export default function MyNav() {
    // ...
    const [activeTab, setActiveTab] = useState<string | null>('tab1');

    return (<>
            <section>
                <div className={styles.adminContainer}>
                    <Tabs variant="outline" value={activeTab} onTabChange={setActiveTab}>
                        <Tabs.List>
                            <Tabs.Tab value={"tab1"}>Tab 1 title text</Tabs.Tab>
                            <Tabs.Tab value={"tab2"}>Tab 2 title text</Tabs.Tab>
                        </Tabs.List>

                        <Tabs.Panel value={"tab1"}>
                            {activeTab==='tab1' ? <MyTab1Component/> : <></>}
                        </Tabs.Panel>
                        <Tabs.Panel value={"tab2"}>
                            {activeTab==='tab2' ? <MyTab2Component someprop={'yay'}/> : <></>}
                        </Tabs.Panel>
                    </Tabs>
                </div>
            </section>
        </>
    );
}
`

**Reply by pkro** [reply]
I just saw that 5.2 now supports a `keepMounted` prop on Tabs - thanks!
